hack/tools/logcheck/pkg/logcheck.go

```diff
@@ -0,0 +1,489 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pkg
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/token"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"go/types"
+	"golang.org/x/exp/utf8string"
+	"golang.org/x/tools/go/analysis"
+)
+
+const (
+	unstructuredCheck = "unstructured"
+	parametersCheck   = "parameters"
+	klogCheck         = "klog"
+	withHelpersCheck  = "with-helpers"
+)
+
+type checks map[string]*bool
+
+type config struct {
+	enabled       checks
+	fileOverrides RegexpFilter
+}
+
+func (c config) isEnabled(check string, filename string) bool {
+	return c.fileOverrides.Enabled(check, *c.enabled[check], filename)
+}
+
+// Analyser creates a new logcheck analyser.
+func Analyser() *analysis.Analyzer {
+	c := config{
+		enabled: checks{
+			unstructuredCheck: new(bool),
+			parametersCheck:   new(bool),
+			klogCheck:         new(bool),
+			withHelpersCheck:  new(bool),
+		},
+	}
+	c.fileOverrides.validChecks = map[string]bool{}
+	for key := range c.enabled {
+		c.fileOverrides.validChecks[key] = true
+	}
+	logcheckFlags := flag.NewFlagSet("", flag.ExitOnError)
+	prefix := "check-"
+	logcheckFlags.BoolVar(c.enabled[unstructuredCheck], prefix+unstructuredCheck, true, `When true, logcheck will warn about calls to unstructured
+klog methods (Info, Infof, Error, Errorf, Warningf, etc).`)
+	logcheckFlags.BoolVar(c.enabled[parametersCheck], prefix+parametersCheck, true, `When true, logcheck will check parameters of structured logging calls.`)
+	logcheckFlags.BoolVar(c.enabled[klogCheck], prefix+klogCheck, false, `When true, logcheck will warn about any legacy klog calls.`)
+	logcheckFlags.BoolVar(c.enabled[withHelpersCheck], prefix+withHelpersCheck, false, `When true, logcheck will warn about direct calls to WithName, WithValues and NewContext.`)
+	logcheckFlags.Var(&c.fileOverrides, "config", `A file which overrides the global settings for checks on a per-file basis via regular expressions.`)
+
+	// Use env variables as defaults. This is necessary when used as plugin
+	// for golangci-lint because of
+	// https://github.com/golangci/golangci-lint/issues/1512.
+	for key, enabled := range c.enabled {
+		envVarName := "LOGCHECK_" + strings.ToUpper(strings.ReplaceAll(string(key), "-", "_"))
+		if value, ok := os.LookupEnv(envVarName); ok {
+			v, err := strconv.ParseBool(value)
+			if err != nil {
+				panic(fmt.Errorf("%s=%q: %v", envVarName, value, err))
+			}
+			*enabled = v
+		}
+	}
+	if value, ok := os.LookupEnv("LOGCHECK_CONFIG"); ok {
+		if err := c.fileOverrides.Set(value); err != nil {
+			panic(fmt.Errorf("LOGCHECK_CONFIG=%q: %v", value, err))
+		}
+	}
+
+	return &analysis.Analyzer{
+		Name: "logcheck",
+		Doc:  "Tool to check use of unstructured logging patterns.",
```

Comment:
Is this still valid? We can now run logcheck with "unstructured" check disabled.

Reply:
Good catch. Updated.